### ISSUE TITLE
builder/amazon: assume_role in shared credentials file not honored

### ISSUE THREAD

**Issue by delfick** [hashicorp/packer#3019]
Hello,

I have a situation where I use Packer on my bamboo agents to build an AMI.

My bamboo agents are setup so that they have no permission other than the ability to assume a role and then we make it assume deployer roles for doing anything with amazon.

This works fine in that I assume a role that can create the AMI and I call packer.

The problem is that building this AMI can take longer than an hour (because reasons) and so what happens is it gets to the point where it needs to stop the instance to make an AMI and the credentials are no longer valid.

How difficult would it be for me to be able to tell packer what role to assume and packer handles assuming that role (either at every step, or intelligently based on the lifetime of the creds) ??

Thanks

Stephen.

**Comment by cbednarski** [comment]
Sorry about the late response. My understanding is that Packer should already be able to assume a role based on the IAM credentials you supply. This happens in the AWS library so it's transparent to us. Does the credential timeout issue still happen if you let packer assume the role on its own instead of providing it with temporary credentials?

**Comment by delfick** [comment]
I believe you're referring to the magic metadata service (169.254.169.254)
unfortunately I can't use that cause it only has the permission to assume
other roles where I'm running packer. And so I must assume a particular
role first.

**Comment by delfick** [comment]
By the looks of it, you could use http://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/stscreds/AssumeRoleProvider.html to do the hard work and just provide the role_arn option to give to that provider :)

**Comment by secretmike** [comment]
This we be great - we have a similar setup with STS AssumeRole and builds taking longer than an hour. After the hour is up we get `Error waiting for AMI: RequestExpired: Request has expired.`

**Comment by rickard-von-essen** [comment]
Could you give some more details about your setup? 

Do you use sts to get temporary credentials and set the env vars: `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, and `AWS_SESSION_TOKEN` or
do you define a profile in `~/aws/config`?

**Comment by secretmike** [comment]
Hi @rickard-von-essen - yes, I'm using environment variables with `AWS_SESSION_TOKEN`.

I have a small wrapper script which calls STS `AssumeRole`, sets the temporary creds in the environment to `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, and `AWS_SESSION_TOKEN` then calls `packer build`.

This all works great - packer runs using the temporary credentials, with the permissions of the assumed role. It only becomes a problem when a packer build takes over an hour, since the AssumeRole call limits the maximum lifetime of the tokens to 1 hour.

Mainly I would love a way for packer (or via AWS Go SDK) to generate new temporary credentials when required.

We have one set of somewhat permanent credentials that can't do anything except assume a role. We would pass those permanent credentials to packer with the `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` variables, along with some kind of `AWS_ASSUME_ROLE=arn:aws:iam::123456789012:role/packer`

Packer would then use something like the `AssumeRoleProvider` (http://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/stscreds/AssumeRoleProvider.html) to generate new temporary credentials every couple minutes (or every time they expired).

If it's possible to set it up as a profile using `~/.aws/config` that would work ok for me - grateful for any pointers if it's possible!

Thanks!

**Comment by rickard-von-essen** [comment]
Could you try using AWS profiles, if you setup your `~/.aws/config`.

```toml
[profile base]

[profile packer-role]
role_arn = arn:aws:iam::1234567890123:role/packer-builder
source_profile = base
```

and `~/.aws/credentials`
```toml
[base]
aws_access_key_id = <AKIA>
aws_secret_access_key = <SECRETKEY>
```

And the you run packer with:
`AWS_PROFILE=packer-role packer build template.json`

I haven't tried it. But from looking in the code it should refresh your credentials. 

(But it doesn't work if your policies require MFA)

**Comment by rickard-von-essen** [comment]
We should update `aws-go-sdk` to the latest and add support for `assume_role_arn` and update the docs accordingly and with some more examples in `amazon.html`.

**Comment by rickard-von-essen** [comment]
#4613 will add support for Assume Role (with or without MFA). Unfortunately it will not honor `assume_role` from a shared credential file since that is not currently supported in the AWS Go SDK.

**Comment by delfick** [comment]
cool

**Comment by spanktar** [comment]
I can confirm that this works in 1.0.1 as noted: https://github.com/hashicorp/packer/issues/3019#issuecomment-254737838

**Comment by rickard-von-essen** [comment]
Fixed in #4996
